### PR TITLE
[SPARK-17319] [SQL] Move addJar from HiveSessionState to SharedState

### PR DESCRIPTION
**Pull request by gatorsmile**
### What changes were proposed in this pull request?

The added jar is shared by all the sessions, because SparkContext does not support sessions. Thus, it makes more sense to move `addJar` from `HiveSessionState` to `HiveSharedState`

This is another step to remove Hive client usage in `HiveSessionState`.

Different sessions are sharing the same class loader (IsolatedClientLoader), and thus, `metadataHive.addJar(path)` basically loads the JARs for all the sessions. Thus, no impact is made if we move `addJar` from `HiveSessionState` to `HiveExternalCatalog`.
### How was this patch tested?

The existing test cases.

### CODE REVIEW

**Comment by gatorsmile** [comment]
cc @cloud-fan @yhuai

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/InMemoryCatalog.scala [hunk 1]
```diff
@@ -509,4 +509,10 @@ class InMemoryCatalog(
     StringUtils.filterPattern(catalog(db).functions.keysIterator.toSeq, pattern)
   }
 
+  // --------------------------------------------------------------------------
+  // Resources
+  // --------------------------------------------------------------------------
+
+  override def addJar(path: String): Unit = { /* no-op */ }
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/InMemoryCatalog.scala]: Will no-op here make user think that the Jar is loaded?

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/InMemoryCatalog.scala]: Yea, I think throw exception is better.

File: sql/core/src/main/scala/org/apache/spark/sql/internal/SessionState.scala [hunk 2]
```diff
@@ -171,6 +171,7 @@ private[sql] class SessionState(sparkSession: SparkSession) {
   }
 
   def addJar(path: String): Unit = {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/SessionState.scala]: so the `addJar` is not session-scoped by definition? I think `sparkSession.sparkContext.addJar(path)` is also cross-session

**Comment by cloud-fan** [comment]
@gatorsmile is the `ADD JAR` command in hive session-scoped? Our current implementation may be wrong..

File: sql/core/src/main/scala/org/apache/spark/sql/internal/SessionState.scala (hunk 2 above)
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/SessionState.scala]: hmm, I think the addition of resources should be session-scoped.

**Comment by viirya** [comment]
https://cwiki.apache.org/confluence/display/Hive/LanguageManual+Cli#LanguageManualCli-HiveResources

Looks like it is.

**Comment by gatorsmile** [comment]
@cloud-fan @viirya After reading Hive's code, the JAR's scope should be session based. See the code:
https://github.com/apache/hive/blob/0438701395161325a429b4fd8211213276aa0fef/ql/src/java/org/apache/hadoop/hive/ql/session/SessionState.java#L1188-L1200

Let me think how to fix it.

**Comment by gatorsmile** [comment]
After reading the code, it sounds like adding session-scoped ADD JAR is not simple if we want to pass it to every worker node after the PR: https://github.com/apache/spark/pull/8909

CC @davies

**Comment by cloud-fan** [comment]
ok maybe we should update the document and say that `ADD JAR` is cross-session, also move the `addJar` method to `SharedState`?

**Comment by gatorsmile** [comment]
It sounds good to me! Will make a change tomorrow. Thanks!

**Comment by gatorsmile** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala [hunk 3]
```diff
@@ -37,7 +37,7 @@ case class AddJarCommand(path: String) extends RunnableCommand {
   }
 
   override def run(sparkSession: SparkSession): Seq[Row] = {
-    sparkSession.sessionState.addJar(path)
+    sparkSession.sharedState.addJar(path)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala]: we should also update the document of `AddJarCommand`, and mention that it's cross-session

File: sql/core/src/main/scala/org/apache/spark/sql/internal/SessionState.scala [hunk 4]
```diff
@@ -43,6 +40,11 @@ private[sql] class SessionState(sparkSession: SparkSession) {
   // Note: These are all lazy vals because they depend on each other (e.g. conf) and we
   // want subclasses to override some of the fields. Otherwise, we would get a lot of NPEs.
 
+  // Since both sharedState and sessionState are lazily initialized in SparkSession,
+  // we need to ensure sharedState are initialized before the sessionState; otherwise,
+  // hive-site.xml might not be loaded when we using it.
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/SessionState.scala]: is it an existing bug?

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/SessionState.scala]: Before this PR, we did it in an implicit way. : )
 https://github.com/apache/spark/blob/2b10ebe6ac1cdc2c723cb47e4b88cfbf39e0de08/sql/core/src/main/scala/org/apache/spark/sql/internal/SessionState.scala#L155

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala (hunk 3 above)
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala]: Sure, will do it.

File: sql/core/src/main/scala/org/apache/spark/sql/internal/SessionState.scala (hunk 4 above)
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/SessionState.scala]: ah i see, then can we just make `sharedState` not a lazy val?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala [hunk 5]
```diff
@@ -37,13 +38,13 @@ case class AddJarCommand(path: String) extends RunnableCommand {
...
     Seq(Row(0))
   }
 }
 
 /**
- * Adds a file to the current session so it can be used.
+ * Adds a cross-session file so it can be used.
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala]: Also updated the command of `ADD FILE`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalog.scala [hunk 6]
```diff
@@ -184,4 +184,9 @@ abstract class ExternalCatalog {
 
   def listFunctions(db: String, pattern: String): Seq[String]
 
+  // --------------------------------------------------------------------------
+  // Resources
+  // --------------------------------------------------------------------------
+
+  def addJar(path: String): Unit
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalog.scala]: I'm thinking of how to define the semantic of `ExternalCatalog.addJar`, any ideas?

Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalog.scala]: Add a resource to the underlying Hive metastore for DDL operations

For example, if we do not use HiveClient to pass the `ADD JAR` command to Hive metastore, we are unable to create the table. Thus, it sounds fine to put `addJar` into `ExternalCatalog`.
```Scala
    val testJar = TestHive.getHiveFile("hive-hcatalog-core-0.13.1.jar").getCanonicalPath
    sql(s"ADD JAR $testJar")
    sql(
      """
        |CREATE TABLE t1(a string, b string)
        |ROW FORMAT SERDE 'org.apache.hive.hcatalog.data.JsonSerDe'
      """.stripMargin)
```

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalog.scala]: Do we have to mention hive here? I'd like to add some documents here to describe the semantic, which can explain why `InMemoryCatalog` can do nothing but `HiveExternalCatalog` need some extra logic

Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalog.scala]: yea, I don't think we should limit `addJar` semantics to Hive.

Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalog.scala]: Add a jar resource to the underlying external catalog system for DDL operations. And followed by the example of Hive.

Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalog.scala]: Let me rephrase it. 

>Add a JAR resource to the underlying external catalog for DDL (e.g. CREATE TABLE) and DML (e.g., LOAD TABLE) operations.

>For example, when users create a Hive serde table, they can specify a custom Serializer-Deserializer (SerDe) class. When Hive metastore is unable to access the custom SerDe JAR (e.g., not on the Hive classpath), the JAR file must be added at runtime using this API.

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalog.scala]: LGTM

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalog.scala]: this also implies that `InMemoryCatalog` can't work if users specify a custom SerDe class in CREATE TABLE. Considering this, should we throw exception in `InMemoryCatalog.addJar`?

Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalog.scala]: I think so.

Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalog.scala]: Let me change that. Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala [hunk 7]
```diff
@@ -69,6 +71,29 @@ private[sql] class SharedState(val sparkContext: SparkContext) extends Logging {
   val jarClassLoader = new NonClosableMutableURLClassLoader(
     org.apache.spark.util.Utils.getContextOrSparkClassLoader)
 
+  /**
+   * Add a global-scoped jar
+   */
+  def addJar(path: String): Unit = {
+    if (sparkContext.conf.get(CATALOG_IMPLEMENTATION) == "hive") {
+      // Hive metastore supports custom serde.
+      externalCatalog.addJar(path)
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala]: We call `addJar` only when the external catalog is `Hive metastore`. This is just consistent with our current implementation: 

https://github.com/apache/spark/blob/261c55dd8808502fb7f3384eb537d26a4a8123d7/sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionState.scala#L104

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala]: when will we call `SharedState.addJar`? only the `ADD JAR` command?

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala]: Also used by run-time function loading when Hive support is enabled. 

If Hive support is not enabled, `ADD JAR` is the only caller of `SharedState.addJar`. It is just a SQL interface of `sparkContext.addJar`. I am not sure whether end users are using `ADD JAR` for `sparkContext.addJar`. Thus, I am hesitant to make any external behavior change in this PR.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalog.scala [hunk 8]
```diff
@@ -184,4 +184,17 @@ abstract class ExternalCatalog {
...
+
+  /**
+   * Add a JAR resource to the underlying external catalog for DDL (e.g. CREATE TABLE) and DML
+   * (e.g., LOAD TABLE) operations.
+   *
+   * For example, when users create a Hive serde table, they can specify a custom
+   * Serializer-Deserializer (SerDe) class. When Hive metastore is unable to access the custom SerDe
+   * JAR (e.g., not on the Hive classpath), the JAR file must be added at runtime using this API.
+   */
+  def addJar(path: String): Unit
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalog.scala]: I'm still a little hesitant to add this concept. After we consolidate hive and data source tables, do we still need it? data source resolution is done before we save table metadata to metastore.

Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalog.scala]: Yeah, we still need it after hive and data source tables are consolidated. The major reason is for supporting custom Hive Serde, custom Hive file format, and custom UDF/UDAF. They are major features for the existing Hive users. These supports might be critical for migration from Hive to Spark.

Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalog.scala]: Let me copy and paste the reasons why users need the customization capabilities:
- When to add a new File Format?

> User has files with special file formats not supported by Hive yet, and users don’t want to convert the files before loading into Hive.
> User has a more efficient way of storing data on disk.
- When to add a new File Format?

- When to add a new SerDe?

> User has data with special serialized format not supported by Hive yet, and users don’t want to convert the data before loading into Hive.
> User has a more efficient way of serializing the data on disk.

Also, the existing UDF, UDAF and UDTF can be reused without any rewriting.

Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalog.scala]: This is a good point. Once hive is consolidated into data source table, we can specify jar in data source way e.g., option.

Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalog.scala]: We still need an `addJar` API in the `ExternalCatalog` for passing it to the underlying Hive metastore, even if we can specify jar in data source options. Here, we are not storing the `Jar` info in the metastore. The Hive metastore needs it for creating/loading the tables.

Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalog.scala]: After hive is consolidated into data source, custom serde should be the business of data source, not `ExternalCatalog`, right? And it is possibly that the external catalog is not hive metastore.

However, it might be too far to bring the consideration here.

Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalog.scala]: Everything is doable. Basically, we need to provide a NATIVE support for reading/writing custom/built-in Hive serde/file-format tables, instead of using `HiveClient`. Ideally, we should implement our own write path for Hive serde tables, instead of calling HiveClient's loadTable/loadPartition APIs, and MORE. After all these changes, we can completely get rid of Hive metastore and HiveClient APIs. 

However, we have to consider the priority, code stability, and maintainability. All the above efforts/supports are for migration from legacy Hive to Spark. For the Spark users who do not have legacy Hive system, they should use our built-in file-format sources and data source APIs, instead of writing/using custom Hive serde/file format.

Comment (jiangxb1987) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalog.scala]: In my opinion, making addJar as a data source option is a bit more complex than our current way, respect the `addJar` API. I suggest we back @gatorsmile 's choice and move forward, since it keeps most of the prior behavior.
